test/unit/ui_utils_spec.js

```diff
@@ -398,4 +399,236 @@ describe('ui_utils', function() {
+    function scrollOverDocument(pages, horizontally = false) {
+      const size = pages.reduce(function(max, { div, }) {
+        return Math.max(
+          max,
+          horizontally ?
+            div.offsetLeft + div.clientLeft + div.clientWidth :
+            div.offsetTop + div.clientTop + div.clientHeight);
+      }, 0);
+      for (let i = 0; i < size; i += 7) {
+        for (let j = i + 5; j < size; j += (j - i)) {
+          const scroll = horizontally ? {
+            scrollTop: 0,
+            scrollLeft: i,
+            clientHeight: 10000,
+            clientWidth: j - 1,
+          } : {
+            scrollTop: i,
+            scrollLeft: 0,
+            clientHeight: j - i,
+            clientWidth: 10000,
+          };
+          expect(getVisibleElements(scroll, pages, false, horizontally))
+            .toEqual(slowGetVisibleElements(scroll, pages));
+        }
+      }
+    }
+
+    it('with pages of varying height', function() {
+      const pages = makePages([
+        [[50, 20], [20, 50]],
+        [[30, 12], [12, 30]],
+        [[20, 50], [50, 20]],
+        [[50, 20], [20, 50]],
+      ]);
+      scrollOverDocument(pages);
+    });
+
+    it('widescreen challenge', function() {
+      const pages = makePages([
+        [[10, 50], [10, 60], [10, 70], [10, 80], [10, 90]],
+        [[10, 90], [10, 80], [10, 70], [10, 60], [10, 50]],
+        [[10, 50], [10, 60], [10, 70], [10, 80], [10, 90]],
+      ]);
+      scrollOverDocument(pages);
+    });
+
+    it('works with horizontal scrolling', function() {
+      const pages = makePages([
+        [[10, 50], [20, 20], [30, 10]],
+      ]);
+      scrollOverDocument(pages, true);
+    });
+
+    // This sub-suite is for a notionally internal helper function for
+    // getVisibleElements.
+    describe('backtrackBeforeAllVisibleElements', function() {
+      // Layout elements common to all tests
+      const tallPage = [10, 50];
+      const shortPage = [10, 10];
+
+      // A scroll position that ensures that only the tall pages in the second
+      // row are visible
+      const top1 =
+        20 + SPACING + // height of the first row
+        40; // a value between 30 (so the short pages on the second row are
+            // hidden) and 50 (so the tall pages are visible)
+
+      // A scroll position that ensures that all of the pages in the second row
+      // are visible, but the tall ones are a tiny bit cut off
+      const top2 = 20 + SPACING + // height of the first row
+        10; // a value greater than 0 but less than 30
+
+      // These tests refer to cases enumerated in the comments of
+      // backtrackBeforeAllVisibleElements.
+      it('handles case 1', function() {
+        const pages = makePages([
+          [[10, 20], [10, 20], [10, 20], [10, 20]],
+          [tallPage, shortPage, tallPage, shortPage],
+          [[10, 50], [10, 50], [10, 50], [10, 50]],
+          [[10, 20], [10, 20], [10, 20], [10, 20]],
+          [[10, 20]],
+        ]);
```

Comment:
`pages` is different in cases 1-3. They don't have to be, since we're fudging the result of binary search, but I think it's illustrative to use cases that would actually produce the binary search results that we're asserting.

Reply:
You're right, never mind. I compared case 3 and 4, which are identical, except for the `top` values.